starboard/shared/libiamf/iamf_config_reader.cc

```diff
@@ -0,0 +1,531 @@
+    }                                                                     \
+  } while (0)
+
+// From
+// https://aomediacodec.github.io/iamf/v1.0.0-errata.html#obu-header-syntax.
+constexpr int kObuTypeCodecConfig = 0;
+constexpr int kObuTypeAudioElement = 1;
+constexpr int kObuTypeMixPresentation = 2;
+constexpr int kObuTypeSequenceHeader = 31;
+
+// From
+// https://aomediacodec.github.io/iamf/v1.0.0-errata.html#obu-codecconfig.
+constexpr int kFourccOpus = 0x4f707573;
+constexpr int kFourccIpcm = 0x6970636d;
+
+}  // namespace
+
+IamfConfigReader::BufferReader::BufferReader(const uint8_t* buf, size_t size)
+    : buf_(buf), pos_(0), size_(size), error_(false) {}
+
+bool IamfConfigReader::BufferReader::Read1(uint8_t* ptr) {
+  if (!HasBytes(sizeof(uint8_t)) || !ptr) {
+    error_ = true;
+    return false;
+  }
+  *ptr = buf_[pos_++];
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::Read4(uint32_t* ptr) {
+  if (!HasBytes(sizeof(uint32_t)) || !ptr) {
+    error_ = true;
+    return false;
+  }
+  std::memcpy(ptr, &buf_[pos_], sizeof(uint32_t));
+  *ptr = ByteSwap(*ptr);
+  pos_ += sizeof(uint32_t);
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::ReadLeb128(uint32_t* ptr) {
+  if (!HasBytes(sizeof(uint32_t)) || !ptr) {
+    error_ = true;
+    return false;
+  }
+  int bytes_read = ReadLeb128Internal(buf_ + pos_, ptr);
+  if (bytes_read < 0) {
+    error_ = true;
+    return false;
+  }
+  pos_ += bytes_read;
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::ReadString(std::string* str) {
+  int bytes_read = ReadStringInternal(buf_ + pos_, str);
+  if (bytes_read < 0) {
+    error_ = true;
+    return false;
+  }
+  pos_ += bytes_read;
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::SkipBytes(size_t size) {
+  if (!HasBytes(size)) {
+    error_ = true;
+    return false;
+  }
+  pos_ += size;
+  return true;
+}
+
+bool IamfConfigReader::BufferReader::SkipLeb128() {
+  uint32_t val;
+  return ReadLeb128(&val);
+}
+
+bool IamfConfigReader::BufferReader::SkipString() {
+  std::string str;
+  return ReadString(&str);
+}
+
+int IamfConfigReader::BufferReader::ReadLeb128Internal(const uint8_t* buf,
+                                                       uint32_t* value) {
+  SB_DCHECK(buf);
+  SB_DCHECK(value);
+  *value = 0;
+  bool error = true;
+  size_t i = 0;
+  for (; i < sizeof(uint32_t); ++i) {
+    uint8_t byte = buf[i];
+    *value |= ((byte & 0x7f) << (i * 7));
+    if (!(byte & 0x80)) {
+      error = false;
+      break;
+    }
+  }
+
+  if (error) {
+    return -1;
+  }
+  return i + 1;
+}
+
+int IamfConfigReader::BufferReader::ReadStringInternal(const uint8_t* buf,
+                                                       std::string* str) {
+  SB_DCHECK(buf);
+
+  int remaining_size = static_cast<int>(size_) - pos_;
+  if (remaining_size <= 0) {
+    return -1;
+  }
+
+  // The size of the string is capped to 128 bytes.
+  int max_str_size = std::min(remaining_size, 128);
```

Comment:
nit: it's better to declare the max length as a const variable. As it's only used in this function, we can declare the const in this function.